### PR TITLE
Variables: Limit rendering of options in dropdown to improve search performance 

### PR DESCRIPTION
**Pull request by jonny-gg**
**What this PR does / why we need it**:
Solve the first dropdown box when the data volume is large, keep within 1000/ In the case of a large amount of data, such as 20,000, the delay of the first render drop-down box will be longer

**Which issue(s) this PR fixes**:

Fixes #
Solve the first dropdown box when the data volume is large, keep within 1000
**Special notes for your reviewer**:

### CODE REVIEW

**Comment by zhukaiqiang** [comment]
I have the same problem, I hope to merge this PR and solve it as soon as possible

**Comment by torkelo** [comment]
Think this is an ok temporary  solution until we add virtualized rendering. 

@hugohaggmark can you take a look and see if this is ok change to limit rendering of options in the dropdown?

File: public/app/features/variables/pickers/OptionsPicker/reducer.ts [hunk 1]
```diff
@@ -78,6 +78,10 @@ const updateOptions = (state: OptionsPickerState): OptionsPickerState => {
 
     return { ...option, selected };
   });
+  if (Array.isArray(state.options) && state.options.length > OPTIONS_LIMIT) {
+    const options = state.options;
+    state.options = options.slice(0, OPTIONS_LIMIT);
+  }
```
Comment (hugohaggmark) [inline public/app/features/variables/pickers/OptionsPicker/reducer.ts:84]: Thank you for contributing @guoqn this looks good but I think that you could change this to call `applyLimit` function instead and remove other calls to applyLimit as `updateOptions` is called on all the reducer calls.

Thank you
Also, we need some tests for this new logic. 
```suggestion
  state.options = applyLimit(state.options);
```

**Review by hugohaggmark — CHANGES_REQUESTED** [review]
This looks good but needs some tests and refactor.

File: public/app/features/variables/pickers/OptionsPicker/reducer.ts (hunk 1 above)
Comment (guoqn) [reply public/app/features/variables/pickers/OptionsPicker/reducer.ts:84]: Wow, this looks a little bit more neat

Comment (guoqn) [reply public/app/features/variables/pickers/OptionsPicker/reducer.ts]: but, the other two USES of 'applyLimit' have their own logic and are not recommended

**Review by jonny-gg — COMMENTED** [review]
A PR was resubmitted, using 'applyLimit' function @hugohaggmark  thx

**Comment by hugohaggmark** [comment]
> A PR was resubmitted, using 'applyLimit' function @hugohaggmark thx

Great work, there where other places where applyLimit was called that should be removed like
https://github.com/grafana/grafana/blob/fe66b2173f136b344b3041b974dcc52c9687f0ef/public/app/features/variables/pickers/OptionsPicker/reducer.ts#L243
https://github.com/grafana/grafana/blob/fe66b2173f136b344b3041b974dcc52c9687f0ef/public/app/features/variables/pickers/OptionsPicker/reducer.ts#L249

I think those should be removed as well as they call updateOptions too @guoqn

**Comment by guoqn** [comment]
> > A PR was resubmitted, using 'applyLimit' function @hugohaggmark thx
> 
> Great work, there where other places where applyLimit was called that should be removed like
> https://github.com/grafana/grafana/blob/fe66b2173f136b344b3041b974dcc52c9687f0ef/public/app/features/variables/pickers/OptionsPicker/reducer.ts#L243
> 
> 
> https://github.com/grafana/grafana/blob/fe66b2173f136b344b3041b974dcc52c9687f0ef/public/app/features/variables/pickers/OptionsPicker/reducer.ts#L249
> 
> I think those should be removed as well as they call updateOptions too @guoqn

如果逻辑中没有错误，其他两个调用方将过滤原始数据。如果删除，第一次过滤的内容将作为第二次搜索的依据，这将导致过滤函数失效

**Comment by guoqn** [comment]
> > > A PR was resubmitted, using 'applyLimit' function @hugohaggmark thx
> > 
> > 
> > Great work, there where other places where applyLimit was called that should be removed like
> > https://github.com/grafana/grafana/blob/fe66b2173f136b344b3041b974dcc52c9687f0ef/public/app/features/variables/pickers/OptionsPicker/reducer.ts#L243
> > 
> > https://github.com/grafana/grafana/blob/fe66b2173f136b344b3041b974dcc52c9687f0ef/public/app/features/variables/pickers/OptionsPicker/reducer.ts#L249
> > 
> > I think those should be removed as well as they call updateOptions too @guoqn
> 
> If there are no errors in the logic, the other two calls to `applyLimit` are used to filter the source data (**action.payload**). If deleted, the contents of the first filter will be used as the basis for the second search, which will invalidate the filter function @hugohaggmark

**Comment by guoqn** [comment]
不好意思，点错了pull request

**Comment by guoqn** [comment]
> > A PR was resubmitted, using 'applyLimit' function @hugohaggmark thx
> 
> Great work, there where other places where applyLimit was called that should be removed like
> https://github.com/grafana/grafana/blob/fe66b2173f136b344b3041b974dcc52c9687f0ef/public/app/features/variables/pickers/OptionsPicker/reducer.ts#L243
> 
> 
> https://github.com/grafana/grafana/blob/fe66b2173f136b344b3041b974dcc52c9687f0ef/public/app/features/variables/pickers/OptionsPicker/reducer.ts#L249
> 
> I think those should be removed as well as they call updateOptions too @guoqn



> > A PR was resubmitted, using 'applyLimit' function @hugohaggmark thx
> 
> Great work, there where other places where applyLimit was called that should be removed like
> https://github.com/grafana/grafana/blob/fe66b2173f136b344b3041b974dcc52c9687f0ef/public/app/features/variables/pickers/OptionsPicker/reducer.ts#L243
> 
> 
> https://github.com/grafana/grafana/blob/fe66b2173f136b344b3041b974dcc52c9687f0ef/public/app/features/variables/pickers/OptionsPicker/reducer.ts#L249
> 
> I think those should be removed as well as they call updateOptions too @guoqn



> there where other places where applyLimit was called that should be removed like

The payload is restricted on the raw data (`action.Payload`) and the rendered data (`state.options`), unless you modify the `updateOptions` parameters, which can be risky

**Comment by hugohaggmark** [comment]
Now I'm superconfused, you propose a change to the `updateOptions` function right? Both lines that I added to my comment are also calling updateOptions, but before they do that they reduce the amount of items in the state with `applyLimit`.

This is what I suggest:

```ts
    updateOptionsAndFilter: (state, action: PayloadAction<VariableOption[]>): OptionsPickerState => {
      const searchQuery = trim((state.queryValue ?? '').toLowerCase());

      state.options = action.payload.filter(option => {
        const text = Array.isArray(option.text) ? option.text.toString() : option.text;
        return text.toLowerCase().indexOf(searchQuery) !== -1;
      });

      state.highlightIndex = 0;

      return applyStateChanges(state, updateDefaultSelection, updateOptions);
    },
    updateOptionsFromSearch: (state, action: PayloadAction<VariableOption[]>): OptionsPickerState => {
      state.options = action.payload;
      state.highlightIndex = 0;

      return applyStateChanges(state, updateDefaultSelection, updateOptions);
    },
```

**Comment by guoqn** [comment]
> Now I'm superconfused, you propose a change to the `updateOptions` function right? Both lines that I added to my comment are also calling updateOptions, but before they do that they reduce the amount of items in the state with `applyLimit`.
> 
> This is what I suggest:
> 
> ```ts
>     updateOptionsAndFilter: (state, action: PayloadAction<VariableOption[]>): OptionsPickerState => {
>       const searchQuery = trim((state.queryValue ?? '').toLowerCase());
> 
>       state.options = action.payload.filter(option => {
>         const text = Array.isArray(option.text) ? option.text.toString() : option.text;
>         return text.toLowerCase().indexOf(searchQuery) !== -1;
>       });
> 
>       state.highlightIndex = 0;
> 
>       return applyStateChanges(state, updateDefaultSelection, updateOptions);
>     },
>     updateOptionsFromSearch: (state, action: PayloadAction<VariableOption[]>): OptionsPickerState => {
>       state.options = action.payload;
>       state.highlightIndex = 0;
> 
>       return applyStateChanges(state, updateDefaultSelection, updateOptions);
>     },
> ```

e.... it's my bad ,sorry for everything before.

**Comment by hugohaggmark** [comment]
> > Now I'm superconfused, you propose a change to the `updateOptions` function right? Both lines that I added to my comment are also calling updateOptions, but before they do that they reduce the amount of items in the state with `applyLimit`.
> > This is what I suggest:
> > ```ts
> >     updateOptionsAndFilter: (state, action: PayloadAction<VariableOption[]>): OptionsPickerState => {
> >       const searchQuery = trim((state.queryValue ?? '').toLowerCase());
> > 
> >       state.options = action.payload.filter(option => {
> >         const text = Array.isArray(option.text) ? option.text.toString() : option.text;
> >         return text.toLowerCase().indexOf(searchQuery) !== -1;
> >       });
> > 
> >       state.highlightIndex = 0;
> > 
> >       return applyStateChanges(state, updateDefaultSelection, updateOptions);
> >     },
> >     updateOptionsFromSearch: (state, action: PayloadAction<VariableOption[]>): OptionsPickerState => {
> >       state.options = action.payload;
> >       state.highlightIndex = 0;
> > 
> >       return applyStateChanges(state, updateDefaultSelection, updateOptions);
> >     },
> > ```
> 
> e.... it's my bad ,sorry for everything before.

No worries, would be outstanding if you could add a test for this new logic as well.

**Comment by guoqn** [comment]
> 太好了

will try it.

**Comment by guoqn** [comment]
> > > Now I'm superconfused, you propose a change to the `updateOptions` function right? Both lines that I added to my comment are also calling updateOptions, but before they do that they reduce the amount of items in the state with `applyLimit`.
> > > This is what I suggest:
> > > ```ts
> > >     updateOptionsAndFilter: (state, action: PayloadAction<VariableOption[]>): OptionsPickerState => {
> > >       const searchQuery = trim((state.queryValue ?? '').toLowerCase());
> > > 
> > >       state.options = action.payload.filter(option => {
> > >         const text = Array.isArray(option.text) ? option.text.toString() : option.text;
> > >         return text.toLowerCase().indexOf(searchQuery) !== -1;
> > >       });
> > > 
> > >       state.highlightIndex = 0;
> > > 
> > >       return applyStateChanges(state, updateDefaultSelection, updateOptions);
> > >     },
> > >     updateOptionsFromSearch: (state, action: PayloadAction<VariableOption[]>): OptionsPickerState => {
> > >       state.options = action.payload;
> > >       state.highlightIndex = 0;
> > > 
> > >       return applyStateChanges(state, updateDefaultSelection, updateOptions);
> > >     },
> > > ```
> > 
> > 
> > e.... it's my bad ,sorry for everything before.
> 
> No worries, would be outstanding if you could add a test for this new logic as well.

will try it. The test section is my weakness.

**Comment by guoqn** [comment]
Added test cases for updateOptionsFromSearch and updateOptionsFromSearch when data volumes are high

**Comment by hugohaggmark** [comment]
> Added test cases for `updateOptionsFromSearch` and `updateOptionsFromSearch` when data volumes are high @hugohaggmark

The tests look outstanding @guoqn, before we can merge this we need just one test for the `showOptions` action with the applyLimit logic. Let me know if I can help out.

**Comment by guoqn** [comment]
> > Added test cases for `updateOptionsFromSearch` and `updateOptionsFromSearch` when data volumes are high @hugohaggmark
> 
> The tests look outstanding @guoqn, before we can merge this we need just one test for the `showOptions` action with the applyLimit logic. Let me know if I can help out.

ok, i can help, wait a moment

**Comment by guoqn** [comment]
> > Added test cases for `updateOptionsFromSearch` and `updateOptionsFromSearch` when data volumes are high @hugohaggmark
> 
> The tests look outstanding @guoqn, before we can merge this we need just one test for the `showOptions` action with the applyLimit logic. Let me know if I can help out.

one test for the showOptions action with the applyLimit logic has commit @hugohaggmark

File: public/app/features/variables/pickers/OptionsPicker/reducer.test.ts [hunk 2]
```diff
@@ -911,4 +911,35 @@ describe('optionsPickerReducer', () => {
...
+      }
+
+      reducerTester<OptionsPickerState>()
+        .givenReducer(optionsPickerReducer, cloneDeep(initialState))
+        .whenActionIsDispatched(showOptions(payload))
+        .thenStateShouldEqual({
+          ...initialState,
+          options: checkOptions,
+          id: payload.id,
+          multi: payload.multi,
```
Comment (hugohaggmark) [inline public/app/features/variables/pickers/OptionsPicker/reducer.test.ts:940]: Great work on this test! Just one small nit and then I think we can merge this.
```suggestion
          id: '0',
          multi: false,
```

Comment (guoqn) [reply public/app/features/variables/pickers/OptionsPicker/reducer.test.ts]: OK, fixed。

**Review by hugohaggmark — APPROVED** [review]
This has been such a great contribution @guoqn, thank you 🌟

**Comment by guoqn** [comment]
Thank you very much for your patient guidance.I'll explore virtual lists to improve on this temporary alternative

**Pull request merged**